src/data/num/to_bits.lean

```diff
@@ -0,0 +1,150 @@
+For convenience, the decoding function always returns a value. If the most significant bit isn't
+`1`, it implicitly changes it to `1` when decoding. We then prove several lemmas about this
+encoding.
+-/
+
+namespace pos_num
+
+/-- Convert a `pos_num` to a `list bool` in little-endian form (with LSB at the head).
+This is an equivalence, so the msb (which is always present for positive numbers) is omitted. -/
+def to_trailing_bits : pos_num → list bool
+| 1 := []
+| (bit0 xs) := ff :: xs.to_trailing_bits
+| (bit1 xs) := tt :: xs.to_trailing_bits
+
+/-- Equivalence between `pos_num` and `list bool` with lsb at the head and msb omitted -/
+def equiv_list_bool : pos_num ≃ list bool :=
+{ to_fun := to_trailing_bits,
+  inv_fun := λ l, l.foldr (λ b n, (cond b bit1 bit0) n) 1,
+  left_inv := λ n, by induction n; simpa [to_trailing_bits],
+  right_inv := λ l, by { induction l with hd, { refl, }, cases hd; simpa [to_trailing_bits], } }
+
+@[simp] lemma equiv_list_bool_one : equiv_list_bool 1 = [] := rfl
+@[simp] lemma equiv_list_bool_bit0 (n : pos_num) :
+  equiv_list_bool (bit0 n) = ff :: equiv_list_bool n := rfl
+@[simp] lemma equiv_list_bool_bit1 (n : pos_num) :
+  equiv_list_bool (bit1 n) = tt :: equiv_list_bool n := rfl
+@[simp] lemma equiv_list_bool_symm_cons_tt (l : list bool) :
+  equiv_list_bool.symm (tt :: l) = bit1 (equiv_list_bool.symm l) := rfl
+@[simp] lemma equiv_list_bool_symm_cons_ff (l : list bool) :
+  equiv_list_bool.symm (ff :: l) = bit0 (equiv_list_bool.symm l) := rfl
+
+lemma to_trailing_bits_len (n : pos_num) : (equiv_list_bool n).length = nat.log 2 n :=
+begin
+  induction n with b ih b ih, { erw equiv_list_bool_one, simp, },
+  { simp [cast_bit1, nat.bit1_val b, ih], },
+  { simp [cast_bit0, nat.bit0_val b, ih], }
+end
+
+lemma equiv_list_bool_symm_lt (ls : list bool) : (equiv_list_bool.symm ls : ℕ) < 2^(ls.length+1) :=
+begin
+  induction ls with hd tl ih, { simp [equiv_list_bool], },
+  cases hd,
+  { simp only [equiv_list_bool_symm_cons_ff, pos_num.cast_bit0],
+    rw [nat.bit0_val, mul_comm, pow_add], simpa, },
+  { simp only [equiv_list_bool_symm_cons_tt, pos_num.cast_bit1],
+    rw [nat.bit1_val, add_comm, pow_add],
+    apply @nat.lt_of_div_lt_div _ _ 2,
+    simpa [nat.add_mul_div_left, nat.div_eq_of_lt], },
+end
+
+end pos_num
+
+namespace num
+
+/-- An encoding function of the binary numbers in bool. -/
+def to_bits : num → list bool
+| num.zero := []
+| (num.pos n) := (pos_num.equiv_list_bool n) ++ [tt]
+
+@[simp] lemma to_bits_zero : to_bits 0 = [] := rfl
+
+@[simp] lemma to_bits_bit0 (n : num) (hn : n ≠ 0) : (num.bit0 n).to_bits = ff :: n.to_bits :=
+by { cases n, { contradiction, }, simp [to_bits, num.bit0], }
+
+@[simp] lemma to_bits_bit1 (n : num) : (num.bit1 n).to_bits = tt :: n.to_bits :=
+by { cases n, { refl, }, simp [to_bits, num.bit1], }
+
+/-- A decoding function from `list bool` to `num` -/
+def of_bits (L : list bool) : num := if L = [] then 0 else pos_num.equiv_list_bool.symm L.init
```

Comment:
This function does not handle lists with trailing ff correctly. It should be defined directly by recursion

Reply:
How do you think trailing `ff` should be handled? If it is implicitly putting a `tt` at the end, then the main disadvantage of that is it breaks monotonicity.